multistream-2.0/spec.md

```diff
@@ -0,0 +1,403 @@
+# Multistream 2.0
+
+This proposal describes a replacement protocol for multistream-select.
+
+## Protocols
+
+This document proposes 6 new, micro-protocols with two guiding principles:
+
+1. Composition over complexity.
+2. Every byte and round-trip counts.
+
+This document *does not*, in fact, propose a protocol *negotiation* protocol.
+Instead, it proposes a set of stream/protocol management protocols that can be
+composed to flexibly negotiate protocols.
+
+First, this document proposes 4 protocol "negotiation" protocols. "Negotiation"
+is in quotes because none of these protocols actually involve negotiating
+anything.
+
+1. `multistream/advertise`: Inform the remote end about which protocols we speak
+   and. This should partially replace the current identify protocol.
+2. `multistream/use`: Selects the stream's protocol using a multicodec.
+3. `multistream/dynamic`: Selects the stream's protocol using a string protocol name.
+4. `multistream/contextual`: Selects the stream's protocol using a protocol ID
+   defined by the *receiver*, valid for the duration of the "session"
+   (underlying connection). To use this, the *receiver* must have used the
+   `multistream/advertise` To inform the initiator of *it's* mapping between
+   protocols and contextual IDs.
+
+Second, this document proposes 2 auxiliary protocols that can be used with the 4
+multistream protocols to actually negotiate protocols. These are *primarily*
+useful (a) in packet-based protocols (without sessions) and (b) when initially
+negotiating a transport session (before protocols have been advertised and the
+stream multiplexer has been configured).
+
+1. `serial-stream`: A simple stream "multiplexer" that can multiplex multiple
+   streams *in serial* over the same connection. That is, it allows us to
+   negotiate a protocol, use it, and then return to multistream. It also allows
+   us to speculatively choose a single protocol and then drop back down to
+   multistream if that doesn't work.
+2. `speculative-stream`: A speculative stream "multiplexer" where the initiator
+   can speculatively initiate multiple streams and the receiver must select at
+   most one and discard the others.
```

Comment:
presumably there is  a mechanism to inform the peer of the actual selection.

Reply:
Yes (described in the protocol description)